### ISSUE TITLE
🤦VS Code extensions do not support esm

### ISSUE THREAD

**Issue by Adam-it** [pnp/vscode-viva#114]
## 🎯 Aim

The aim of this issue is to track the support of es modules in VS Code extensions.
Issues with more info:
- https://github.com/microsoft/vscode/issues/130367

Currently using CLI for Microsoft 365 directly from PnP builds is not possible due to that, so for now as a temporary workaround the CLI dependency is added from a local version of the package which was reverted to common js
https://github.com/Adam-it/cli-microsoft365/tree/cli-cjs

**Comment by Malix-off** [comment]
Yep, that's a serious disappointment.
Did you try using [vite](https://vitejs.dev/)?

**Comment by Adam-it** [comment]
> Yep, that's a serious disappointment.
> Did you try using [vite](https://vitejs.dev/)?

Not sure how would that help for a dependency that is not used in webview but in the extension itself 🤔. Do you have any example I could check 👍?

**Comment by Malix-off** [comment]
> > Yep, that's a serious disappointment.
> > Did you try using [vite](https://vitejs.dev/)?
> 
> Not sure how would that help for a dependency that is not used in webview but in the extension itself 🤔. Do you have any example I could check 👍?

Yep, I use vite to bundle my VSCode extension
It's WIP, but the bundling part works; it bundles from my typescript source written in ESNext to a CJS build
Here is the source code: https://github.com/Malix-off/VSC-Extension-StashFlow

**Comment by Adam-it** [comment]
Awesome. I need to check it for sure 🤩. Will it work also if one of the dependencies I use is already build in esm? As this is actually the problem I'm having

**Comment by Malix-off** [comment]
> Will it work also if one of the dependencies I use is already build in esm?
As this is actually the problem I'm having

Yes, it's the main point of bundler, fixing the dumbass cjs vs esm fight and optimize code output with minification and circular resolution, and library imports with tree-shaking
There is theoretically no reason vite couldn't fix your problem in one way or another, I believe

**Comment by Adam-it** [comment]
this is awesome. I need to do a deep dive on this in the coming weekend. 
Thanks @Malix-off for sharing. You Rock 🤩

**Comment by Malix-off** [comment]
You're welcome ❤️

**Comment by estruyf** [comment]
@Malix-off, it is a great tip, also started using Vite a while ago, but now stepping away from it again as Vite dropped CJS support in v5 and will be completely removed in v6.

> More information: https://vitejs.dev/guide/migration.html#deprecate-cjs-node-api

I stumbled on this issue as I had some issues with types that were incorrectly referenced during the build and made the extension fail.

It would be better to look at `esbuild` or `tsup` to be future-proof.

**Comment by Malix-off** [comment]
> Vite dropped CJS support in v5

Nope, they dropped [their own CJS Node API](https://vitejs.dev/guide/migration.html#deprecate-cjs-node-api), which is not a bad move imho because it would be weird to use a bundle in a non-ESM env

It means that, to use Vite, you need to call it with an ESM syntax, but vite will still be available to output CJS

See https://github.com/Malix-off/VSC-Extension-StashFlow/blob/dev/vite.config.ts as an exemple, the CJS output is a `rollupOption`.

**Comment by Malix-off** [comment]
@estruyf ⬆️

**Comment by Adam-it** [comment]
related issues and PRs
https://github.com/microsoft/vscode-loader/pull/57
https://github.com/microsoft/vscode-loader/issues/36
https://github.com/microsoft/vscode-loader/issues/56

**Comment by Adam-it** [comment]
https://code.visualstudio.com/updates/v1_94#_esm-is-shipping-for-vs-code

**Comment by Adam-it** [comment]
Rechecked this topic and still no change on that note. Seems like we could wait for another year to get this 🫤

**Comment by Malix-Labs** [comment]
Yep.

Related issue : https://github.com/microsoft/vscode/issues/130367

**Comment by Adam-it** [comment]
https://github.com/[jrieken/vscode-esm-sample-extension](https://github.com/jrieken/vscode-esm-sample-extension)

**Comment by Adam-it** [comment]
> Nice!
> 
> @Adam-it what is missing for that issue to be closed as resolved?

I guess we need to update SPFx Toolkit to esm and also use esm dependencies 😉

**Comment by Adam-it** [comment]
lets give this a recheck 
https://github.com/microsoft/vscode/issues/130367
